File: story-engine/app/natural_language/terrain_phrases.py
# ============================================================================
# Terrain phrases (pure lookup helpers)
# ----------------------------------------------------------------------------
# Port of the pure functions from
# backend/domains/story/services/naturalLanguage/terrainPhrases.js.
#
# loadTerrainPhrases (the DB-reading half of the original file) stays in Node:
# it is called from the Game domain during day generation, before narrateDay
# runs, and its result is already embedded in `day.terrain_phrases` by the
# time this service receives the payload. Nothing here talks to a database.
# ============================================================================
import random


def pick_phrase(phrases_map, region_name, category, rng=random.random):
    """Pick a random phrase for a region/category combination."""
    phrases = (phrases_map or {}).get(region_name, {}).get(category)
    if not isinstance(phrases, list) or len(phrases) == 0:
        return None
    return phrases[int(rng() * len(phrases))]


def pick_phrase_for_regions(phrases_map, region_names, category, rng=random.random):
    """Pick a phrase from the first region in the list that has one for the category."""
    if not isinstance(region_names, list):
        return None
    for name in region_names:
        phrase = pick_phrase(phrases_map, name, category, rng)
        if phrase:
            return phrase
    return None

File: backend/domains/story/services/naturalLanguage/terrainPhrases.js
// ============================================================================
// Terrain phrases service
// ----------------------------------------------------------------------------
// Loads region-specific narrative phrases for terrain/road categories from the
// region_biome_descriptions table. Falls back to null when no phrase exists,
// letting callers keep their default descriptions.
//
// NOTE: the pure lookup helpers that used to live here (pickPhrase,
// pickPhraseForRegions) moved to the story-engine Python port
// (story-engine/app/natural_language/terrain_phrases.py), since prompt
// assembly now happens there. This file keeps only the DB read, which the
// Game domain still needs during day generation (domains/game/services/world/
// tripDay.js, via domains/game/adapters/storyClient.js).
// ============================================================================

import pool from '../../../../db.js';

/**
 * Build a lookup map of region-specific phrases.
 * @param {string[]} regionNames
 * @param {string[]} [categories]
 * @returns {Promise<Object>} { regionName: { category: [phrase, ...] } }
 */
export async function loadTerrainPhrases(regionNames, categories = []) {
  if (!Array.isArray(regionNames) || regionNames.length === 0) {
    return {};
  }

  const categoryFilter =
    Array.isArray(categories) && categories.length > 0
      ? 'AND category = ANY($2::text[])'
      : '';
  const params = categories.length > 0 ? [regionNames, categories] : [regionNames];

  try {
    const { rows } = await pool.query(
      `SELECT region_name, category, phrases
       FROM region_biome_descriptions
       WHERE region_name = ANY($1::text[])
       ${categoryFilter}`,
      params
    );

    const map = {};
    for (const row of rows) {
      if (!map[row.region_name]) {
        map[row.region_name] = {};
      }
      map[row.region_name][row.category] = row.phrases || [];
    }
    return map;
  } catch (err) {
    // If the table does not exist (or any other DB error), return an empty map
    // so the day can still be generated using fallback descriptions.
    if (err.code === '42P01') {
      console.warn('⚠️ region_biome_descriptions table not found; using fallback terrain descriptions.');
    } else {
      console.warn('⚠️ Could not load terrain phrases:', err.message);
    }
    return {};
  }
}
